README.rst
MAC Authorisation Module for Django
===================================

:Description: A basic implementation of the RFC draft oauth-v2-http-mac-01__ for Django
:Author:      Nicholas Devenish

.. __: http://tools.ietf.org/html/draft-ietf-oauth-v2-http-mac-01

Overview
--------

This package implements the current draft of the OAuth-2-related `MAC authorisation protocol`__. This is designed to allow cryptographically reliable requests to be made over a connection that is susceptible to interception. It is based on a shared "secret" which is used to calculate a cryptographic signature at both client and resource server to verify that the request was authentic.

.. __: http://tools.ietf.org/html/draft-ietf-oauth-v2-http-mac-01

This implementation very simply allows a django view to require that this authentication method be used, or to optionally use the authentication information to determine the user of the request.

Features
--------
These are the features and features of the protocol that this package provides:

* Unique nonce/timestamp/client ID checking; this prevents the possibility of replay attacks (though, see `Limitations`_)
* Uses the hmac-sha-1 algorithm
* partial `ext` header support
* Uses existing Django User framework
* Allows optional usage of credentials
* Authentication errors are communicated back in the WWW-Authentication error parameter
* Tested on Django 1.3 and 1.4

Usage
-----

After adding ``auth_mac`` to your project's settings, and syncing your db, you can create a new set of Credentials through the admin interface, or by instantiating the ``Credentials`` object with an associated user::

  from auth_mac import Credentials
  new_auth = Credentials(user=some_user)
  new_auth.save()

The credentials object will by default be instantiated with a random identifier and secret key, and will have an expiry date set to a day in the future. All of these can be overridden by setting the ``identifier``, ``key`` and ``expiry`` model fields.

When a request is made, you can ensure that the client has authenticated properly by using one of two decorators. The first decorator, **require_credentials**, returns a 401 Unauthorized response if the authentication fails::

  from django.http import HttpResponse
  from auth_mac.decorators import require_credentials

  @require_credentials
  def some_protected_view(request):
    return HttpResponse("This view only runs when authorised!")

And the second decorator, **read_credentials**, allows access even without credentials, but passing credentials will override the ``HttpRequest``'s user property to the authorised user::

  from django.http import HttpResponse
  from auth_mac.decorators import read_credentials

  @read_credentials
  def some_optionally_protected_view(request):
    return HttpResponse("This view can accept Anonymous Users!")

In this second case, if the user is accessing through some other authorisation method i.e. signed in via a session cookie, the credential information (if passed) will overwrite the previous login information.

Limitations
-----------

This is only a very basic implementation of the protocol. Specifically, it does not provide:

* Any way to distribute the secret information. You could do this via an OAuth2 implementation, or manual distribution of the keys. This is because the current design intent is only to provide REST access to a couple of authorised personal clients.
* nonce expiry. At this point in time, nonce values have to be kept indefinitely. Some of the timestamp infrastructure is in place, but until it is completed it is unsafe to remove nonces.
* `hmac-sha-256` is not yet supported
* No rate limitation for protection against flooded requests
* The `ext` parameter is used in calculating the base string, but is not currently handled or tested properly.
* Timestamp verification (including saving of the client offset) is not currently limited. This will be configurable in the future.

CHANGES.txt
c0.1.2, 2012-02-16 --
  * Fixed a bug in Signature instantiation
  * If the port is passed in the HTTP Host header, it is ignored

v0.1.1, 2012-02-15 -- 
  * Remove dependency on South - not needed at the moment
  * Added timezone support when needed to work with Django 1.4
  * Made the dependency on Django 1.3+ explicit

v0.1.0, 2012-02-14 -- Initial release.

auth_mac/migrations/0002_auto__chg_field_nonce_timestamp__chg_field_credentials_identifier__chg.py
# encoding: utf-8
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models

class Migration(SchemaMigration):

    def forwards(self, orm):
        
        # Changing field 'Nonce.timestamp'
        db.alter_column('auth_mac_nonce', 'timestamp', self.gf('django.db.models.fields.DateTimeField')())

        # Changing field 'Credentials.identifier'
        db.alter_column('auth_mac_credentials', 'identifier', self.gf('django.db.models.fields.CharField')(max_length=16))

        # Changing field 'Credentials.key'
        db.alter_column('auth_mac_credentials', 'key', self.gf('django.db.models.fields.CharField')(max_length=16))


    def backwards(self, orm):
        
        # Changing field 'Nonce.timestamp'
        db.alter_column('auth_mac_nonce', 'timestamp', self.gf('django.db.models.fields.DateTimeField')(auto_now_add=True))

        # Changing field 'Credentials.identifier'
        db.alter_column('auth_mac_credentials', 'identifier', self.gf('django.db.models.fields.CharField')(max_length=16, null=True))

        # Changing field 'Credentials.key'
        db.alter_column('auth_mac_credentials', 'key', self.gf('django.db.models.fields.CharField')(max_length=16, null=True))


    models = {
        'auth.group': {
            'Meta': {'object_name': 'Group'},
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '80'}),
            'permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'})
        },
        'auth.permission': {
            'Meta': {'ordering': "('content_type__app_label', 'content_type__model', 'codename')", 'unique_together': "(('content_type', 'codename'),)", 'object_name': 'Permission'},
            'codename': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'content_type': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['contenttypes.ContentType']"}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '50'})
        },
        'auth.user': {
            'Meta': {'object_name': 'User'},
            'date_joined': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'email': ('django.db.models.fields.EmailField', [], {'max_length': '75', 'blank': 'True'}),
            'first_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'groups': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Group']", 'symmetrical': 'False', 'blank': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_active': ('django.db.models.fields.BooleanField', [], {'default': 'True'}),
            'is_staff': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'is_superuser': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'last_login': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'last_name': ('django.db.models.fields.CharField', [], {'max_length': '30', 'blank': 'True'}),
            'password': ('django.db.models.fields.CharField', [], {'max_length': '128'}),
            'user_permissions': ('django.db.models.fields.related.ManyToManyField', [], {'to': "orm['auth.Permission']", 'symmetrical': 'False', 'blank': 'True'}),
            'username': ('django.db.models.fields.CharField', [], {'unique': 'True', 'max_length': '30'})
        },
        'auth_mac.credentials': {
            'Meta': {'object_name': 'Credentials'},
            'expiry': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime(2012, 2, 15, 2, 55, 43, 577482)'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'identifier': ('django.db.models.fields.CharField', [], {'default': "'KDEYV6hWcFtHX8bY'", 'max_length': '16'}),
            'key': ('django.db.models.fields.CharField', [], {'default': "'V9A5j4Y8qSWFY66A'", 'max_length': '16'}),
            'user': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['auth.User']"})
        },
        'auth_mac.nonce': {
            'Meta': {'object_name': 'Nonce'},
            'credentials': ('django.db.models.fields.related.ForeignKey', [], {'to': "orm['auth_mac.Credentials']"}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'nonce': ('django.db.models.fields.CharField', [], {'max_length': '16', 'null': 'True', 'blank': 'True'}),
            'timestamp': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.utcnow'})
        },
        'contenttypes.contenttype': {
            'Meta': {'ordering': "('name',)", 'unique_together': "(('app_label', 'model'),)", 'object_name': 'ContentType', 'db_table': "'django_content_type'"},
            'app_label': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'model': ('django.db.models.fields.CharField', [], {'max_length': '100'}),
            'name': ('django.db.models.fields.CharField', [], {'max_length': '100'})
        }
    }

    complete_apps = ['auth_mac']
